meerk40t/gui/icons.py

```diff
@@ -569,216 +586,207 @@ def GetBitmap(
         def color_id():
             return "--" if color is None else f"{color.red}-{color.green}-{color.blue}"
 
-        def my_id():
-            res_fill = ""
-            for e in self.list_fill:
-                res_fill += str(hash(e)) + ","
-            res_stroke = ""
-            for e in self.list_stroke:
-                res_stroke += str(hash(e[2])) + ","
-
-            return res_fill + "|" + res_stroke
-
-        cache_id = f"{my_id()}|{color_id()}|{resize}|{force_darkmode}"
+        cache_id = f"{self._prehash}|{color_id()}|{resize}|{force_darkmode}"
         if cache_id in _CACHE:
             # print(f"Cache Hit for {cache_id}")
             return _CACHE[cache_id]
         # wincol = wx.SystemSettings().GetColour(wx.SYS_COLOUR_WINDOW)
-        wincol = self._background.GetColour()
-        bmp = wx.Bitmap.FromRGBA(
-            final_icon_width,
-            final_icon_height,
-            wincol.red,
-            wincol.blue,
-            wincol.green,
-            0,
-        )
-        dc = wx.MemoryDC()
-        dc.SelectObject(bmp)
-        # dc.SetBackground(self._background)
-        # dc.SetBackground(wx.RED_BRUSH)
-        # dc.Clear()
-        gc = wx.GraphicsContext.Create(dc)
-        gc.dc = dc
-        stroke_paths = []
-        fill_paths = []
-        # Establish the box...
-        min_x = min_y = max_x = max_y = None
-
-        def get_color(info_color, info_bright, default_color):
-            color = None
-            if info_color is not None:
-                try:
-                    color = wx.Colour(info_color)
-                except AttributeError:
-                    pass
-            if info_bright is not None:
-                if color is None:
-                    color = default_color
-                # brightness is a percentage values below 100 indicate darker
-                # values beyond 100 indicate lighter
-                # no change = 100
-                # What about black? This is a special case, so we only consider
-                ialpha = info_bright / 100.0
-                if color.red == color.green == color.blue == 0 and info_bright > 100:
-                    ialpha = (info_bright - 100) / 100.0
-                    cr = int(255 * ialpha)
-                    cg = int(255 * ialpha)
-                    cb = int(255 * ialpha)
+        with self._lock:
```

Comment:
* This is could still duplicate the work rather than use the cache. You sort of need to do double gated syntax here. 
* The function is also getting a lot complex and maybe should be subdivided into different functions.
* Especially if you could avoid the local caching scheme with the lru_cache decorator.  (https://realpython.com/lru-cache-python/)

Reply:
I'm a bit of a never nester so I would sort of prefer if within the lock it just calls the function that does the rest of the work stuff. You can also avoid the massive ----, ++++ in the diff log that way. And make it a bit easier to see what's being actually altered.

Reply:
The reason this could still double execute is that if two threads hit the lock around the same time, only one thread will get through the lock. Then the other thread will be forced to wait until that thread finishes. When the other thread finishes it will then release this thread which will then repeat all the same work again pointlessly even though it's already done.

Weirdly the correct method is double-gated. You check the cache, then you enter the lock, then you check the cache again. It would seem like you already know the cache was a miss, but you really don't because it could have been a while since you acquired the lock. 

So you do:
```python
        if cache_id in _CACHE:
            return _CACHE[cache_id]
        with self._lock:
            if cache_id in _CACHE:
                return _CACHE[cache_id]
            self._function_that_makes_bitmap_and_fills_cache(cache_id)
```

Reply:
- The double cache check should no longer be relevant, as it has been replaced by the lru-cache decorator.
- Routines have been isolated